feat: Add info panel item `panel` to load and display data on demand

vardhan0604 opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Issue Description

Closes: #2610 

### Approach


### TODOs before merging

Commit:
added changes

vardhan0604 commented:
https://github.com/user-attachments/assets/5abdae0d-4f2d-4632-833f-63c2b6c6139b

Hey @mtrezza, I have added a subset element that allows loading another Cloud Code function on demand. I’m not sure about the styles; could you please let me know what changes and features are required after reviewing the current progress?

Commit:
fix lint

mtrezza commented:
That is amazing! Just what I had in mind. From a first look I things all that's needed is to refine the design of the section. Let me come up with a mock up... Nice work! Maybe @404-html wants to give this a quick review as well.

404-html commented:
Hi again Gents! 👋
> Maybe @404-html wants to give this a quick review as well.

With pleasure!

404-html commented on src/components/AggregationPanel/AggregationPanelComponents.js:
```diff
@@ -95,3 +97,116 @@ export const ButtonElement = ({ item, showNote }) => {
+  };
+
+  const handleToggle = async () => {
+    if ((!isExpanded && !panelData)) {
```
```suggestion
    if (!isExpanded && !panelData) {
```

404-html commented on src/components/AggregationPanel/AggregationPanelComponents.js:
```diff
@@ -95,3 +97,116 @@ export const ButtonElement = ({ item, showNote }) => {
+  const [isLoading, setIsLoading] = useState(false);
+  const [panelData, setPanelData] = useState(null);
+
+  const fetchPanelData = async () => {
```
How about wrapping `fetchPanelData`, `handleToggle` and `handleRefresh` functions into `useCallback`?

404-html commented on src/components/AggregationPanel/AggregationPanelComponents.js:
```diff
@@ -95,3 +97,116 @@ export const ButtonElement = ({ item, showNote }) => {
   );
 };
+
+export const PanelElement = ({ item, showNote, objectId, depth = 0 }) => {
```
Instead of introducing another `PanelElement` I would suggest to extend existing one (in `AggregationPanel.js`), and adopt it to loading data subset and being used recessively. This way we should avoid code duplication.

vardhan0604 replied:
I have tried doing in the latest commits, can you please review it?

Commit:
update

Commit:
fix lint

vardhan0604 commented:
Hi @404-html,  I have updated the code as requested. Could you please review it?

404-html commented:
> Hi @404-html, I have updated the code as requested. Could you please review it?

Let me wait for Manuel's UI suggestions, and after they will be addressed I will do another review.

mtrezza commented:
Not sure why, but for some reason I cannot get the "Show panel" button to show, despite running the same setup like in the previous PR where it worked. @vardhan0604 could you try to rebuild it with the latest commit here and see whether it still works for you?

vardhan0604 commented:
> Not sure why, but for some reason I cannot get the "Show panel" button to show, despite running the same setup like in the previous PR where it worked. @vardhan0604 could you try to rebuild it with the latest commit here and see whether it still works for you?

Yeah, it is working for me. I rebuilt it with the latest commit, and the 'Show panel' button appears as expected. @mtrezza, could you check if parse-dashboard-config.json is configured properly?

mtrezza commented:
Still trying to get this to work. Meanwhile I found this:

https://github.com/parse-community/parse-dashboard/blob/6c4d392a90e46330f22b6232758317be3b546923/src/components/Toolbar/Toolbar.react.js#L18

Only remotely related to this PR, but maybe we can clean it up at once. Why was `classwiseCloudFunctions` added to `Stats`? If I understand it correctly, `classwiseCloudFunctions` is just for the info panel, but Stats is a completely different feature?

mtrezza commented:
https://github.com/parse-community/parse-dashboard/issues/2623 is the reason why the "Show Panel" button isn't showing. Could you take a look?

mtrezza commented on src/dashboard/Data/Browser/DataBrowser.react.js:
```diff
@@ -591,7 +591,7 @@ export default class DataBrowser extends React.Component {
               width={this.state.panelWidth}
               height={Infinity}
-              minConstraints={[100, Infinity]}
+              minConstraints={[400, Infinity]}
```
Please revert if not required for this PR; the panel should be resizable to min. 100 px width.

mtrezza commented on src/dashboard/Data/Browser/DataBrowser.react.js:
```diff
@@ -45,7 +45,7 @@ export default class DataBrowser extends React.Component {
       selectedData: [],
       prevClassName: props.className,
-      panelWidth: 300,
+      panelWidth: 400,
```
Why this change?

vardhan0604 replied:
Sorry, both of them were unintentional changes, i will revert them back in next commit.

mtrezza commented:
Here's a suggestion for a slight UI refinement:

Collapsed:

<img width="432" alt="image" src="https://github.com/user-attachments/assets/c7e18ad1-3319-4e93-8525-29eacc0ec826">

Expanded:

<img width="432" alt="image" src="https://github.com/user-attachments/assets/193b5178-d5f0-45d0-9ef7-464ac60027dc">

### Notes

- I've tried to better distinguish the expandable section from a segment header by using a different background color, but defining colors may be a future feature, so maybe it's better to avoid assuming certain color combinations but rather focus on form. Anyway, here's how it would look like in a less-intrusive gray for example:
  
  <img width="398" alt="image" src="https://github.com/user-attachments/assets/d87d0724-06dc-47c0-a635-37e77d305321">

- When expanded, the sub-panel is encompassed by a 1px gray line to the left and bottom, to visualize the information hierarchy in the panel.
- As we can see in the expanded state, there are 2 "Purchases" headlines right below each other, 1 from the expandable sub-panel item, 1 from the title of the segment within the sub-panel. In cases where this is not desired, the title "Purchases" of the first segment can be avoided by simply not specifying a segment title. That feature doesn't exist yet as the `title` field is a mandatory property of a segment, but it will be easy to add in a future PR.
- I have used the same design language for expandable sections as we already do in the data browser classes for saved filters: 
  - Collapsed:
    <img width="247" alt="image" src="https://github.com/user-attachments/assets/b30806be-b00e-4782-9e2a-08b328ed126b">

  - Expanded:
    <img width="250" alt="image" src="https://github.com/user-attachments/assets/b7de0b82-56f9-4308-87c9-3b2d30ba7687">

vardhan0604 commented:
> #2623 is the reason why the "Show Panel" button isn't showing. Could you take a look?

Yeah sure. I will take a look

vardhan0604 commented:
> #2623 is the reason why the "Show Panel" button isn't showing. Could you take a look?

@mtrezza, should I push the changes here or open a new PR? I’ve already fixed it locally and also added the UI changes.

mtrezza commented:
New PR please. It's a fix unrelated to this new feature.

mtrezza commented:
New PR please

Commit:
added changes

vardhan0604 commented:
> New PR please. It's a fix unrelated to this new feature.

@mtrezza can you review it, I have opened the PR #2627 

Also, I’ve implemented the changes according to the UI provided and pushed the changes in this PR. Could you please review and let me know if any additional adjustments are needed.

Commit:
undo panel width size changess

mtrezza committed:
Merge branch 'alpha' into feature/2610

mtrezza commented:
@vardhan0604 Thanks, could you please take a look at https://github.com/parse-community/parse-dashboard/pull/2627#issuecomment-2597228125? I would prefer to first merge that PR to fix the issue, then we can finally merge this one as well.

mtrezza committed:
Merge branch 'alpha' into feature/2610

vardhan0604 commented:
> Also, I’ve implemented the changes according to the UI provided and pushed the changes in this PR. Could you please review and let me know if any additional adjustments are needed.

Hey @mtrezza ,
Since PR #2627 has already been resolved, I’d appreciate it if you could review this one.
Thank you!

mtrezza commented:
Sure, I'll take a look at https://github.com/parse-community/parse-dashboard/pull/2622 first in a bit.

mtrezza committed:
Merge branch 'alpha' into feature/2610

mtrezza committed:
Merge branch 'alpha' into feature/2610

mtrezza committed:
Merge branch 'alpha' into feature/2610

mtrezza commented:
Well done!

The only minor issues I've discovered:

a) The horizontal indentation increases with every sub panel, but should stay the same:

<img width="797" alt="image" src="https://github.com/user-attachments/assets/247e7ee7-eeb6-4f10-92b0-e25c47f583f1" />

b) The reload icon is gray instead of white:
<img width="47" alt="image" src="https://github.com/user-attachments/assets/4b98de5e-fd52-4e3e-8795-c97537a6403d" />

vardhan0604 committed:
fixed horizontal left margin

vardhan0604 committed:
Merge branch 'feature/2610' of https://github.com/vardhan0604/parse-dashboard into feature/2610

vardhan0604 commented:
Hey @mtrezza, I have fixed the requested changes.
This is how it looks now.
![image](https://github.com/user-attachments/assets/d87901df-01fd-4f2f-99fe-643e867c5be4)

mtrezza reviewed (approved):
Looks good! Nice work

parseplatformorg commented:
🎉 This change has been released in version [6.0.0-alpha.25](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-alpha.25)

mtrezza commented:
@vardhan0604 We forgot to add the new item type to the README info [panel section](https://github.com/parse-community/parse-dashboard?tab=readme-ov-file#button-item). Would you mind opening a quick PR for that?

mtrezza commented:
I've added the docs with https://github.com/parse-community/parse-dashboard/pull/2646, no TODO left.

vardhan0604 commented:
> I've added the docs with #2646, no TODO left.

Great, thanks for handling that!

parseplatformorg commented:
🎉 This change has been released in version [6.1.0](https://github.com/parse-community/parse-dashboard/releases/tag/6.1.0)
